File: graph.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <queue>
#include <vector>
#include <climits>
#define MAX 20
#define MAX_VALUE 9999
using namespace std;
struct Queue
{
	int data[MAX];
	int head;
	int tail;
};

void initQueue(Queue &q)
{
	q.head = -1;
	q.tail = -1;
}

bool isEmptyQueue(Queue q)
{
	return q.head == -1;
}

bool isFullQueue(Queue q)
{
	return q.tail == MAX - 1;
}

void enqueue(Queue &q, int u)
{
	if (isFullQueue(q))
	{
		cout << "Queue is full!" << endl;
		return;
	}

	if (isEmptyQueue(q)) {
		q.head = 0;
	}
	q.tail = (q.tail + 1) % MAX;
	q.data[q.tail] = u;
}

int dequeue(Queue &q)
{
	if(isEmptyQueue(q)) {
		cout << "Queue is empty!" << endl;
		return -1;
	}

	int x = q.data[q.head];
	if (q.head == q.tail) {
		q.head = q.tail = -1;
	} else {
		q.head = (q.head + 1) % MAX;
	}
	return x;
}

struct Stack {
	int data[MAX];
	int size;
};

void init(Stack &s) {
	s.size = -1;
}

void push(Stack &s, int u) {
	s.data[++s.size] = u;  // Increment size first
}

int pop(Stack &s) {
	return s.data[s.size--];  // Return the value and then decrement size
}

int top(Stack s) {
	return s.data[s.size];
}

bool isEmpty(Stack s) {
	return s.size == -1;
}

bool isFull(Stack s) {
	return s.size == MAX - 1;
}

struct List {
	int data[MAX];
	int size;
};

void initList(List &l) {
	l.size = 0;
}

List initVisited(List &l, int vertexCount) {
	l.size = vertexCount;
	for (int i = 0; i < vertexCount; i++) {
		l.data[i] = 0;  // Initialize visited as false (0)
	}
	return l;
}

void addList(List &l, int u) {
	l.data[l.size++] = u;
}

void addVisited(List &l, int u) {
	l.data[u] = 1;  // Mark the node as visited
}

void printList(List l) {
	for (int i = 0; i < l.size; i++) {
		cout << l.data[i] << " ";
	}
	cout << endl;
}

int popList(List &l) {
	int u = l.data[l.size - 1];
	l.size--;
	return u;
}

int getVisitedStatus(List l, int u) {
	return l.data[u];  // Get the visited status
}

struct Graph {
	int vertex;
	int matrix[MAX][MAX];
};

void initGraph(Graph &g) {
	for (int i = 0; i < g.vertex; i++) {
		for (int j = 0; j < g.vertex; j++) {
			g.matrix[i][j] = 0;  // Initialize all edges to 0 (no connection)
		}
	}
}

void initWeightedGraph(Graph &g) {
	for (int i = 0; i < g.vertex; i++) {
		for (int j = 0; j < g.vertex; j++) {
			g.matrix[i][j] = -1;  // Initialize all edges to -1 (no connection)
		}
	}
}

int adjacent(Graph g, int u, int v) {
	return g.matrix[u][v];
}

void addEdge(Graph &g, int u, int v) {
	g.matrix[u][v] = 1;
	g.matrix[v][u] = 1;  // Add undirected edge
}

void printGraph(Graph g) {
	for (int i = 0; i < g.vertex; i++) {
		for (int j = 0; j < g.vertex; j++) {
			cout << g.matrix[i][j] << " ";
		}
		cout << endl;
	}
}

void createGraph(Graph &g) {
	cout << "Enter number of vertices: ";
	cin >> g.vertex;
	initGraph(g);
}

int degree(Graph g, int u) {
	int deg = 0;
	for (int i = 0; i < g.vertex; i++) {
		if (g.matrix[u][i] == 1) {
			deg++;
		}
	}
	return deg;
}

// Read graph data from a string
void readGraphFromString(const string &graphData, Graph &g) {
	istringstream stream(graphData);
	stream >> g.vertex;  // Read number of vertices
	initGraph(g);

	int u, v;
	while (stream >> u >> v) {
		addEdge(g, u, v);  // Add edges from the string input
	}
}

void readWeightedGraphFromString(const string &graphData, Graph &g) {
	istringstream stream(graphData);
	stream >> g.vertex;  // Read number of vertices
	initWeightedGraph(g);

	int u, v, w;
	while (stream >> u >> v >> w) {
		g.matrix[u][v] = w;  // Add weighted edges from the string input
	}
}

List neighbors(Graph g, int u) {
	List l;
	initList(l);
	for (int i = 0; i < g.vertex; i++) {
		if (g.matrix[u][i] >= 1) {
			addList(l, i);  // Add neighbors to the list
		}
	}
	return l;
}

List DFS(Graph g, int u) {
	List result;
	initList(result);
	List visited;
	initVisited(visited, g.vertex);
	Stack s;
	init(s);
	push(s, u);

	while (!isEmpty(s)) {
		int u = pop(s);
		if (getVisitedStatus(visited, u) == 0) {  // Check if node is unvisited
			addVisited(visited, u);
			addList(result, u);  // Add the node to the result list
			List neighborsList = neighbors(g, u);
			for (int i = 0; i < neighborsList.size; i++) {
				int v = neighborsList.data[i];
				if (getVisitedStatus(visited, v) == 0) {  // If neighbor is unvisited
					push(s, v);
				}
			}
		}
	}

	return result;
}

List dfsRecursiveHelper(Graph g, int u, List &visited, List &result) {
	// Mark the node as visited
	addVisited(visited, u);
	// Add the node to the result list
	addList(result, u);
	// Explore the neighbors
	List neighborsList = neighbors(g, u);

	for (int i = 0; i < neighborsList.size; i++) {
		int v = neighborsList.data[i];
		if (getVisitedStatus(visited, v) == 0) {  // If the neighbor hasn't been visited
			dfsRecursiveHelper(g, v, visited, result);  // Recursively explore the neighbor
		}
	}

	return result;
}

List dfsRecursive(Graph g, int u) {
	List visited;
	initVisited(visited, g.vertex);
	List result;
	initList(result);
	return dfsRecursiveHelper(g, u, visited, result);  // Call the helper function
}

List BFS(Graph g, int u) {
	List result;
	initList(result);
	List visited;
	initVisited(visited, g.vertex);
	Queue q;
	initQueue(q);
	enqueue(q, u);

	while(!isEmptyQueue(q)) {  // Change condition to while(!isEmptyQueue(q))
		int u = dequeue(q);
		if(getVisitedStatus(visited, u) == 0) {
			addVisited(visited, u);
			addList(result, u);
			List neighborsList = neighbors(g, u);
			for(int i = 0; i < neighborsList.size; i++) {
				int v = neighborsList.data[i];
				if(getVisitedStatus(visited, v) == 0) {
					enqueue(q, v);
				}
			}
		}
	}

	return result;
}

// Stack DijKstra(Graph g, int u, int v) {
//     // Mảng lưu khoảng cách từ u đến các đỉnh còn lại, khởi tạo bằng giá trị vô cùng
//     std::vector<int> dist(g.vertex, MAX_VALUE);
//     dist[u] = 0;  // Khoảng cách từ u đến chính nó là 0

//     // Mảng visited để kiểm tra các đỉnh đã được duyệt
//     std::vector<bool> visited(g.vertex, false);

//     // Mảng parent để lưu lại đỉnh trước đó khi đi qua
//     std::vector<int> parent(g.vertex, -1);

//     Stack stack;
//     init(stack);

//     // Duyệt qua tất cả các đỉnh trong đồ thị
//     for (int i = 0; i < g.vertex; i++) {
//         // Tìm đỉnh chưa được duyệt và có khoảng cách ngắn nhất
//         int minDist = MAX_VALUE;
//         int current = -1;
//         for (int j = 0; j < g.vertex; j++) {
//             if (!visited[j] && dist[j] < minDist) {
//                 minDist = dist[j];
//                 current = j;
//             }
//         }

//         // Nếu không còn đỉnh nào để duyệt, dừng
//         if (current == -1) {
//             break;
//         }

//         // Đánh dấu đỉnh này là đã được duyệt
//         visited[current] = true;

//         // Cập nhật khoảng cách cho các đỉnh kề
//         for (int neighbor = 0; neighbor < g.vertex; neighbor++) {
//             // Kiểm tra nếu có cạnh giữa current và neighbor (trọng số khác -1)
//             if (g.matrix[current][neighbor] != -1 && !visited[neighbor]) {
//                 int edgeWeight = g.matrix[current][neighbor];
//                 int newDist = dist[current] + edgeWeight;
//                 if (newDist < dist[neighbor]) {
//                     dist[neighbor] = newDist;
//                     parent[neighbor] = current;  // Cập nhật đường đi
//                 }
//             }
//         }
//     }

//     // Kiểm tra xem có đường đi từ u đến v không
//     if (dist[v] == MAX_VALUE) {
//         cout << "No path from " << u << " to " << v << endl;
//         return stack;  // Nếu không có đường đi, trả về stack rỗng
//     }

//     // Tạo đường đi từ u đến v (phục hồi đường đi sử dụng parent[])
//     int current = v;
//     while (current != -1) {
//         push(stack, current);  // Đưa đỉnh vào stack
//         current = parent[current];  // Đi đến đỉnh trước đó
//     }

//     return stack;
// }

// Dijkstra algorithm to find the shortest path from source to all other vertices
void dijkstra(Graph &g, int source) {
    vector<int> dist(g.vertex, INT_MAX); // distance array
    dist[source] = 0;  // distance to the source is 0
    
    vector<bool> visited(g.vertex, false); // visited array
    
    // Min-heap to store the vertex with the smallest distance
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, source});  // Start with the source vertex

    while (!pq.empty()) {
        int u = pq.top().second;  // Get the vertex with the smallest distance
        pq.pop();
        
        if (visited[u]) continue;  // Skip if the vertex is already processed
        visited[u] = true;

        // Explore the neighbors of vertex u
        for (int v = 0; v < g.vertex; v++) {
            if (g.matrix[u][v] != -1 && !visited[v]) {  // There is a valid edge and v is not visited
                int newDist = dist[u] + g.matrix[u][v];
                if (newDist < dist[v]) {  // Relax the edge if we found a shorter path
                    dist[v] = newDist;
                    pq.push({dist[v], v});
                }
            }
        }
    }
    
    // Output the shortest distances from source to all other vertices
    cout << "Shortest distances from vertex " << source << ":" << endl;
    for (int i = 0; i < g.vertex; i++) {
        if (dist[i] == INT_MAX)
            cout << "Vertex " << i << " is unreachable" << endl;
        else
            cout << "Distance to vertex " << i << " is " << dist[i] << endl;
    }
}

int main() {
    Graph g;
    string graphData = "5\n0 1 10\n0 2 20\n1 2 5\n1 3 50\n2 3 10\n3 4 2"; // Example graph input
    readWeightedGraphFromString(graphData, g);
    
    int source = 0;  // Source vertex
    dijkstra(g, source);
    
    return 0;
}

// int main() {
//     Graph g;
//     string graphData = "4\n0 1 2\n0 2 4\n1 2 1\n1 3 7\n2 3 3\n";  // Đồ thị có trọng số
//     readWeightedGraphFromString(graphData, g);

//     // In đồ thị trọng số
//     cout << "Weighted Graph adjacency matrix:" << endl;
//     printGraph(g);

//     // Chạy Dijkstra từ đỉnh 0 đến đỉnh 3
//     Stack path = DijKstra(g, 0, 3);

//     // In đường đi từ 0 đến 3
//     cout << "Path from 0 to 3: ";
//     while (!isEmpty(path)) {
//         cout << pop(path) << " ";
//     }
//     cout << endl;

//     return 0;
// }

// int main() {
// 	cout << "Hello, World!" << endl;
// 	Graph g;

// 	// Test case string (graphData)
// 	string graphData = "4\n0 1\n0 2\n1 2\n2 3\n";
// 	string WeightedGraph = "4\n0 1 2\n0 2 3\n1 2 4\n2 3 5\n";

// 	// Read graph from string
// 	readGraphFromString(graphData, g);

// 	// Print the graph's adjacency matrix
// 	cout << "Graph adjacency matrix:" << endl;
// 	printGraph(g);

// 	// Example to print the neighbors of vertex 2
// 	cout << "Neighbors of vertex 2: ";
// 	List neighborsList = neighbors(g, 2);
// 	printList(neighborsList);

// 	// Example to calculate the degree of vertex 2
// 	cout << "Degree of vertex 2: " << degree(g, 2) << endl;

// 	// Example to run DFS from vertex 0
// 	cout << "DFS starting from vertex 0: ";
// 	List dfsResult = DFS(g, 0);
// 	printList(dfsResult);

// 	cout << "DFS starting from vertex 0 by Recursive: ";
// 	List dfs = dfsRecursive(g, 0);
// 	printList(dfs);

// 	cout << "BFS starting from vertex 0: ";
// 	List bfs = BFS(g, 0);
// 	printList(bfs);

// 	return 0;
// }

File: graph.java
import java.util.*;

class UnionFind {
	int parent[];
	int rank[];

	UnionFind(int n) {
		parent = new int[n];
		rank = new int[n];
		for (int i = 0; i < n; i++) {
			parent[i] = i;
			rank[i] = 0;
		}
	}

	int find(int u) {
		if (parent[u] == u) {
			return u;
		}
		return parent[u] = find(parent[u]);
	}

	void union(int u, int v) {
		int pu = find(u);
		int pv = find(v);
		if (pu == pv) {
			return;
		}
		if (rank[pu] < rank[pv]) {
			parent[pu] = pv;
		} else if (rank[pu] > rank[pv]) {
			parent[pv] = pu;
		} else {
			parent[pu] = pv;
			rank[pv]++;
		}
	}
}

class Graph {
	static final int MAX = 20;
	static final int MAX_VALUE = 9999;  // Represents no connection (infinity)
	int vertex;
	int[][] matrix = new int[MAX][MAX];

	Graph(int[][] matrix, int vertex) {
		this.vertex = vertex;
		this.matrix = matrix;
	}

	// Initialize a graph with no connections
	void initGraph() {
		for (int i = 0; i < vertex; i++) {
			for (int j = 0; j < vertex; j++) {
				matrix[i][j] = -1; // -1 indicates no edge
			}
		}
	}

	// Initialize weighted graph with maximum values (indicating no connection)
	void initWeightedGraph() {
		for (int i = 0; i < vertex; i++) {
			for (int j = 0; j < vertex; j++) {
				matrix[i][j] = MAX_VALUE; // Initialize all edges to MAX_VALUE (no connection)
			}
		}
	}

	// Add an edge to the graph
	void addEdge(int u, int v, int w, boolean directed) {
		matrix[u][v] = w;
		if (!directed) {
			matrix[v][u] = w; // Add undirected edge
		}
	}

	// Print the graph's adjacency matrix
	void printGraph() {
		for (int i = 0; i < vertex; i++) {
			for (int j = 0; j < vertex; j++) {
				if (matrix[i][j] == MAX_VALUE) {
					System.out.print("\u221E "); // Symbol for unreachable
				} else {
					System.out.print(matrix[i][j] + " ");
				}
			}
			System.out.println();
		}
	}

	void printGraphData(String graphData) {
		System.out.println("Graph data:");
		System.out.println(graphData);
	}

	// Read graph data from a string
	void readWeightedGraphFromString(String graphData) {
		Scanner scanner = new Scanner(graphData);
		vertex = scanner.nextInt();
		initWeightedGraph();

		while (scanner.hasNext()) {
			int u = scanner.nextInt();
			int v = scanner.nextInt();
			int w = scanner.nextInt();
			addEdge(u, v, w, false);
		}
	}

	void BFS(int start) {
		boolean[] visited = new boolean[vertex];
		Queue<Integer> queue = new LinkedList<>();
		visited[start] = true;
		queue.add(start);

		while (!queue.isEmpty()) {
			int u = queue.poll();
			System.out.print(u + " ");
			for (int v = 0; v < vertex; v++) {
				if (matrix[u][v] != -1 && !visited[v]) {
					visited[v] = true;
					queue.add(v);
				}
			}
		}
	}

	void DFS(int start) {
		boolean[] visited = new boolean[vertex];
		Stack<Integer> stack = new Stack<>();
		stack.push(start);

		while (!stack.isEmpty()) {
			int u = stack.pop();
			if (!visited[u]) {
				visited[u] = true;
				System.out.print(u + " ");
				for (int v = 0; v < vertex; v++) {
					if (matrix[u][v] != -1 && !visited[v]) {
						stack.push(v);
					}
				}
			}
		}
	}

	// Dijkstra algorithm to find the shortest path from source vertex
	void dijkstra(int source) {
		int[] dist = new int[vertex];
		Arrays.fill(dist, MAX_VALUE);
		dist[source] = 0;

		int[] path = new int[vertex];
		for (int i = 0; i < vertex; i++) {
			path[i] = -1;
		}
		path[source] = source;

		boolean[] visited = new boolean[vertex];
		PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
		pq.add(new int[] {source, 0}); // Start with the source vertex

		while (!pq.isEmpty()) {
			int u = pq.poll()[0];
			if (visited[u]) continue;
			visited[u] = true;

			for (int v = 0; v < vertex; v++) {
				if (matrix[u][v] != MAX_VALUE && !visited[v]) {
					int newDist = dist[u] + matrix[u][v];
					if (newDist < dist[v]) {
						dist[v] = newDist;
						pq.add(new int[] {v, dist[v]});
						path[v] = u;
					}
				}
			}
		}

		System.out.println("Shortest paths from vertex " + source + ":");
		for (int i = 0; i < vertex; i++) {
			if (dist[i] != MAX_VALUE) {
				System.out.print("Path to " + i + ": ");
				printPath(path, i);
				System.out.println(" with distance: " + dist[i]);
			}
		}
	}

	void bellmanFord(int source) {
		int[] dist = new int[vertex];
		Arrays.fill(dist, MAX_VALUE);
		dist[source] = 0;

		int[] path = new int[vertex];
		for (int i = 0; i < vertex; i++) {
			path[i] = -1;
		}
		path[source] = source;

		for (int i = 0; i < vertex - 1; i++) {
			for(int u = 0; u < vertex; u++) {
				for(int v = 0; v < vertex; v++) {
					if(matrix[u][v] != MAX_VALUE) {
						int newDist = dist[u] + matrix[u][v];
						if(newDist < dist[v]) {
							dist[v] = newDist;
							path[v] = u;
						}
					}
				}
			}
		}

		for(int u = 0; u < vertex; u++) {
			for(int v = 0; v < vertex; v++) {
				if(matrix[u][v] != MAX_VALUE) {
					int newDist = dist[u] + matrix[u][v];
					if(newDist < dist[v]) {
						System.out.println("Graph contains negative cycle");
						return;
					}
				}
			}
		}
	}

	// Helper method to print the path
	private void printPath(int[] path, int vertex) {
		if (vertex == -1) {
			System.out.print("No path"); // If the path is invalid (unreachable)
			return;
		}
		if (path[vertex] == vertex) {
			System.out.print(vertex + " "); // Base case: if vertex is the source itself
			return;
		}
		printPath(path, path[vertex]); // Recursively print the path
		System.out.print(vertex + " ");
	}

	void printMST(int parent[], int graph[][])
	{
		System.out.println("Edge \tWeight");
		for (int i = 1; i < vertex; i++)
			System.out.println(parent[i] + " - " + i + "\t"
			                   + graph[parent[i]][i]);
	}

	void prim(int source) {
		int parent[] = new int [vertex];
		int key[] = new int [vertex];
		Arrays.fill(key, MAX_VALUE);
		key[source] = 0;
		parent[source] = -1;
		boolean[] mstSet = new boolean[vertex];

		PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
		pq.add(new int[] {source, 0});

		while(!pq.isEmpty()) {
			int u = pq.poll()[0];
			mstSet[u] = true;

			for(int v = 0; v < vertex; v++) {
				int weight = matrix[u][v];

				if(weight != MAX_VALUE && !mstSet[v] && weight < key[v]) {
					key[v] = weight;
					parent[v] = u;
					pq.add(new int[] {v, key[v]});
				}
			}
		}

		printMST(parent, matrix);
	}

    // Hàm lấy danh sách các cạnh mà không trùng và sắp xếp theo trọng số tăng dần
    List<int[]> getEgde() {
        Set<String> visited = new HashSet<>();  // Sử dụng Set để đảm bảo không trùng
        List<int[]> edges = new ArrayList<>();  // Lưu các cặp cạnh

        for (int i = 0; i < vertex; i++) {
            for (int j = i + 1; j < vertex; j++) {  // Chỉ kiểm tra một chiều (i, j), không kiểm tra lại (j, i)
                if (matrix[i][j] != MAX_VALUE && matrix[i][j] != -1) {
                    String edgeKey = i + "-" + j;
                    if (!visited.contains(edgeKey)) {
                        visited.add(edgeKey);  // Đánh dấu cặp cạnh đã được duyệt
                        edges.add(new int[]{i, j, matrix[i][j]});
                    }
                }
            }
        }
        
        // Sắp xếp các cạnh theo trọng số tăng dần
        Collections.sort(edges, new Comparator<int[]>() {
            @Override
            public int compare(int[] edge1, int[] edge2) {
                return Integer.compare(edge1[2], edge2[2]); // So sánh trọng số của 2 cạnh
            }
        });

        return edges;
    }

    // In các cạnh ra màn hình
    void prinEgde(List<int[]> edges) {
        for (int[] edge : edges) {
            System.out.println(edge[0] + " " + edge[1] + " " + edge[2]);
        }
    }

    void kruskal() {
        List<int[]> edges = getEgde();
        UnionFind uf = new UnionFind(vertex);
        List<int[]> mst = new ArrayList<>();

        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            int w = edge[2];

            if (uf.find(u) != uf.find(v)) {
                mst.add(new int[]{u, v, w});
                uf.union(u, v);
            }
        }

        System.out.println("MST:");
        for (int[] edge : mst) {
            System.out.println(edge[0] + " " + edge[1] + " " + edge[2]);
        }
    }
}

public class Main {
	public static void main(String[] args) {
		int graph[][] = {
			{ 9999, 2, 9999, 6, 9999 },
			{ 2, 9999, 3, 8, 5 },
			{ 9999, 3, 9999, 9999, 7 },
			{ 6, 8, 9999, 9999, 9 },
			{ 9999, 5, 7, 9, 9999 }
		};
		Graph g = new Graph(graph, graph.length);
		g.printGraph();
		g.prinEgde(g.getEgde());
	}
}
